File: finalyearproject/repos/dkd-oversight/src/Relay.cs
// gov-outbox-relay: at-least-once, ordered, stop-on-first-failure. Headers carry event_id +
// producer_context=government (fleet R6 convention).
using System.Text;
using Confluent.Kafka;
using Npgsql;

namespace OversightSvc;

public sealed class OutboxRelay(NpgsqlDataSource ds, Config cfg, ILogger<OutboxRelay> log)
    : BackgroundService
{
    // ct in Task.Run guards start only; the loop cancels cooperatively via ct checks.
    protected override Task ExecuteAsync(CancellationToken ct) => Task.Run(async () =>
    {
        using var producer = new ProducerBuilder<string, string>(new ProducerConfig
        {
            BootstrapServers = cfg.KafkaBrokers,
            Acks = Acks.All,
            EnableIdempotence = true,
            CompressionType = CompressionType.Snappy,
            AllowAutoCreateTopics = false,
        }).Build();
        while (!ct.IsCancellationRequested)
        {
            try
            {
                await DrainOnce(producer, ct);
            }
            catch (OperationCanceledException) { break; }
            catch (Exception e)
            {
                log.LogError(e, "relay drain failed; retrying");
            }
            try { await Task.Delay(500, ct); } catch (OperationCanceledException) { break; }
        }
        producer.Flush(TimeSpan.FromSeconds(5));
    }, ct);

    private async Task DrainOnce(IProducer<string, string> producer, CancellationToken ct)
    {
        List<Stores.OutboxRow> rows;
        await using (var cx = await ds.OpenConnectionAsync(ct))
            rows = await Stores.FetchUnpublished(cx, 200);
        foreach (var row in rows)
        {
            var msg = new Message<string, string>
            {
                Key = row.Key,
                Value = row.Payload,
                Headers = new Headers
                {
                    { "event_id", Encoding.UTF8.GetBytes(row.EventId) },
                    { "producer_context", "government"u8.ToArray() },
                },
            };
            try
            {
                await producer.ProduceAsync(row.Topic, msg, ct);
            }
            catch (ProduceException<string, string> e)
            {
                log.LogError(e, "publish failed for {topic} — stopping this drain (ordering)", row.Topic);
                break;
            }
            await using var cx = await ds.OpenConnectionAsync(ct);
            await Stores.MarkPublished(cx, row.Id, Domain.NowMs());
        }
    }
}

File: finalyearproject/repos/dkd-fraud/src/fraud/relay.py
"""Outbox relay: at-least-once, ordered, stop-on-first-failure. Headers carry event_id +
producer_context=fraud (fleet R6 convention); downstream inbox dedup absorbs duplicates."""

from __future__ import annotations

import logging
import threading

from confluent_kafka import KafkaException, Producer
from psycopg_pool import ConnectionPool

from fraud import ids, stores

log = logging.getLogger("fraud.relay")


class OutboxRelay:
    def __init__(self, pool: ConnectionPool, brokers: str) -> None:
        self._pool = pool
        self._producer = Producer({
            "bootstrap.servers": brokers,
            "enable.idempotence": True,
            "acks": "all",
            "compression.type": "snappy",
            "allow.auto.create.topics": False,
        })
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None

    def start(self) -> None:
        self._thread = threading.Thread(target=self._run, name="outbox-relay", daemon=True)
        self._thread.start()

    def _run(self) -> None:
        while not self._stop.is_set():
            try:
                self.drain_once()
            except Exception:
                log.exception("relay drain failed; retrying")
            self._stop.wait(0.5)

    def drain_once(self) -> int:
        published = 0
        with self._pool.connection() as cx:
            rows = stores.fetch_unpublished(cx, 200)
            cx.rollback()
        for row in rows:
            try:
                self._produce_sync(row)
            except KafkaException:
                log.exception("publish failed for %s — stopping this drain (ordering)", row.topic)
                break
            with self._pool.connection() as cx:
                stores.mark_published(cx, row.id, ids.now_ms())
                cx.commit()
            published += 1
        return published

    def _produce_sync(self, row: stores.OutboxRow) -> None:
        done = threading.Event()
        holder: list[object] = []

        def cb(err: object, _msg: object) -> None:
            if err is not None:
                holder.append(err)
            done.set()

        self._producer.produce(
            topic=row.topic,
            key=row.partition_key.encode(),
            value=row.payload.encode(),
            headers=[("event_id", row.event_id.encode()),
                     ("producer_context", b"fraud")],
            on_delivery=cb,
        )
        self._producer.flush(10)
        if not done.is_set() or holder:
            raise KafkaException(holder[0] if holder else "delivery not confirmed")

    def close(self) -> None:
        self._stop.set()
        if self._thread is not None:
            self._thread.join(timeout=5)
        self._producer.flush(5)
